internal/dcache/replication_manager/replication_manager.go

```diff
@@ -0,0 +1,213 @@
+		common.Assert(false, fmt.Sprintf("failed to get chunk from node %s for request %+v", targetNodeID, *rpcReq))
+		return nil, err
+	}
+
+	common.Assert(rpcResp != nil, fmt.Sprintf("GetChunk RPC response is nil for request %+v", *rpcReq))
+	common.Assert(rpcResp.Chunk != nil, fmt.Sprintf("chunk in GetChunk RPC response is nil for request %+v", *rpcReq))
+	common.Assert(rpcResp.Chunk.Address != nil, fmt.Sprintf("address of chunk in GetChunk RPC response is nil for request %+v", *rpcReq))
+
+	log.Debug("ReplicationManager::ReadMV: GetChunk RPC response: chunk lmt %v, time taken %v, component RVs %v, chunk address %+v",
+		rpcResp.ChunkWriteTime, rpcResp.TimeTaken, rpcResp.PeerRV, *rpcResp.Chunk.Address)
+
+	req.Data = rpcResp.Chunk.Data
+
+	// TODO: in GetChunk RPC request add data buffer to the request
+	// TODO: in GetChunk RPC response return bytes read
+
+	// TODO: should we validate the hash of the chunk here?
+	hash := getMD5Sum(rpcResp.Chunk.Data)
+	if hash != rpcResp.Chunk.Hash {
+		log.Err("ReplicationManager::ReadMV: Hash mismatch for the chunk read from node %s for request %+v", targetNodeID, *rpcReq)
+		common.Assert(false, fmt.Sprintf("hash mismatch for the chunk read from node %s for request %+v", targetNodeID, *rpcReq))
+		return nil, fmt.Errorf("hash mismatch for the chunk read from node %s for request %+v", targetNodeID, *rpcReq)
+	}
+
+	resp := &ReadMvResponse{
+		// TODO: update this filed after bytes read in response
+		BytesRead: int64(len(rpcResp.Chunk.Data)),
+	}
+
+	return resp, nil
+}
+
+func WriteMV(req *WriteMvRequest) (*WriteMvResponse, error) {
+	if req == nil {
+		log.Err("ReplicationManager::WriteMV: Received nil WriteMV request")
+		common.Assert(false, "received nil WriteMV request")
+		return nil, fmt.Errorf("received nil WriteMV request")
+	}
+
+	log.Debug("ReplicationManager::WriteMV: Received WriteMV request: %v", req.toString())
+
+	if err := req.isValid(); err != nil {
+		log.Err("ReplicationManager::WriteMV: Invalid WriteMV request parameters [%v]", err.Error())
+		common.Assert(false, fmt.Sprintf("invalid WriteMV request parameters [%v]", err.Error()))
+		return nil, err
+	}
+
+	componentRVs := getComponentRVsForMV(req.MvName)
+	if len(componentRVs) == 0 {
+		log.Err("ReplicationManager::WriteMV: No component RVs found for the given MV %s", req.MvName)
+		common.Assert(false, "no component RVs found for the given MV", req.MvName)
+		return nil, fmt.Errorf("no component RVs found for the given MV %s", req.MvName)
+	}
+
+	// check if the mv is online. This is done by checking if all the component RVs are online.
+	// If any of the component RVs are offline, then fail this request.
+	// The caller of WriteMV should make sure that the MV is online at the time calling this function.
+	// This does not ensure that the MV will be online at the time of writing.
+	// If any of the PutChunk RPC calls fail, then the degradeMV method should be called.
```

Comment:
We do allow writing to a degraded or syncing MV, so online MV is not a requirement.
The only requirement is that MV must not be offline (offline MV is one whose all component RVs are offline), writes to offline MV fail rightaway.

Regd'ing RVs, we have the following rules for writing:
- For an "online" and "syncing" RVs we can do PutChunk
- For "offline" and "outofsync" RVs write need to wait till it becomes "syncing". That will happen once the fix-mv and resync-mv workflow is run. So we will need a way to queue these writes and be woken up once those workflows run

Also, if a write to an "online" or "syncing" RV fails with an error signifying RV is offline, then we need to retry the write after running the degrade-mv + fix-mv + resync-mv workflow.